File: sdk/python/src/lmsservice_db/api/certificates/certificates.py

import json
from dataclasses import asdict
from typing import Dict
from faker import Faker
from lmsservice_db.api.certificates.certificates_model import CertificatesCreateModel, CertificatesSearchParams, CertificatesUpdateModel

class Certificates:
    def __init__(self, client):
        self.client = client
        self.faker = Faker()

    def validate_dataclass(self, data: Dict, dataclass_type):
        """
        Validates the input data against the specified dataclass type.
        """
        try:
            instance = dataclass_type(**data)
            return instance
        except TypeError as e:
            raise ValueError(f"Validation failed: {e}")

    def create_certificates(self, data: Dict):
        validated_data = self.validate_dataclass(data, CertificatesCreateModel)
        endpoint = "certificates"
        return self.client.make_request('POST', endpoint, data=asdict(validated_data))

    def get_certificates_by_id(self, certificates_id: str):
        if not isinstance(certificates_id, str):
            raise TypeError(f"Expected 'certificates_id' to be a string, got {type(certificates_id).__name__}.")
        endpoint = f'certificates/{certificates_id}'
        return self.client.make_request('GET', endpoint)

    def search_certificates(self, query_params: Dict):
        validated_data = self.validate_dataclass(query_params, CertificatesSearchParams)
        endpoint = 'certificates/search'
        return self.client.make_request('GET', endpoint, data=asdict(validated_data))

    def update_certificates(self, certificates_id: str, data: Dict):
        if not isinstance(certificates_id, str):
            raise TypeError(f"Expected 'certificates_id' to be a string, got {type(certificates_id).__name__}.")
        validated_data = self.validate_dataclass(data, CertificatesUpdateModel)
        endpoint = f'certificates/{certificates_id}'
        return self.client.make_request('PUT', endpoint, data=asdict(validated_data))

    def delete_certificates(self, certificates_id: str):
        if not isinstance(certificates_id, str):
            raise TypeError(f"Expected 'certificates_id' to be a string, got {type(certificates_id).__name__}.")
        endpoint = f'certificates/{certificates_id}'
        return self.client.make_request('DELETE', endpoint)

    @staticmethod
    async def execute_certificates_operations(client):
        faker = Faker()

        certificates_create_data = {
                        "UserId": "{{USER_ID}}",
            "CertificateNumber": faker.Lorem.Sentence(3),
            "CertificateName": faker.Person.FullName(),
            "CertificateType": faker.Lorem.Sentence(3),
            "IssuedDate": faker.Lorem.Sentence(3),
            "ExpiryDate": faker.Date.Past().ToString("yyyy-MM-ddTHH:mm:ss"),
            "CertificateUrl": faker.Internet.Url(),
            "FinalGrade": faker.Random.Float(1, 100),
            "CreditHours": faker.Random.Int(1, 10000),
            "Skills": faker.Lorem.Sentence(3),
            "IsVerified": faker.Random.Bool(),
            "VerificationUrl": faker.Internet.Url(),
            "IssuedBy": faker.Random.Guid(),
        }

        certificates_update_data = {
                        "UserId": "{{USER_ID}}",
            "CertificateNumber": faker.Lorem.Sentence(3),
            "CertificateName": faker.Person.FullName(),
            "CertificateType": faker.Lorem.Sentence(3),
            "IssuedDate": faker.Lorem.Sentence(3),
            "ExpiryDate": faker.Date.Past().ToString("yyyy-MM-ddTHH:mm:ss"),
            "CertificateUrl": faker.Internet.Url(),
            "FinalGrade": faker.Random.Float(1, 100),
            "CreditHours": faker.Random.Int(1, 10000),
            "Skills": faker.Lorem.Sentence(3),
            "IsVerified": faker.Random.Bool(),
            "VerificationUrl": faker.Internet.Url(),
            "IssuedBy": faker.Random.Guid(),
        }

        search_params = {
                        "UserId": "{{USER_ID}}",
            "CertificateNumber": faker.Lorem.Sentence(3),
            "CertificateName": faker.Person.FullName(),
            "CertificateType": faker.Lorem.Sentence(3),
            "IssuedDate": faker.Lorem.Sentence(3),
            "ExpiryDate": faker.Date.Past().ToString("yyyy-MM-ddTHH:mm:ss"),
            "CertificateUrl": faker.Internet.Url(),
            "FinalGrade": faker.Random.Float(1, 100),
            "CreditHours": faker.Random.Int(1, 10000),
            "Skills": faker.Lorem.Sentence(3),
            "IsVerified": faker.Random.Bool(),
            "VerificationUrl": faker.Internet.Url(),
            "IssuedBy": faker.Random.Guid(),
        }

        new_user = client.certificates.create_certificates(certificates_create_data)
        print('Create:', json.dumps(new_user, indent=2))
        print("\n")

        certificates_id = new_user['Data']['id']
        user = client.certificates.get_certificates_by_id(certificates_id)
        print('GetById:', json.dumps(user, indent=2))
        print("\n")

        search_results = client.certificates.search_certificates(search_params)
        print('Search:', json.dumps(search_results, indent=2))
        print("\n")

        updated_user = client.certificates.update_certificates(certificates_id, certificates_update_data)
        print('Update:', json.dumps(updated_user, indent=2))
        print("\n")

        deleted_user = client.certificates.delete_certificates(certificates_id)
        print('Delete:', json.dumps(deleted_user, indent=2))
        print("\n")

File: sdk/dotnet/lmsservice_db/src/api/certificates/Certificates.cs

using Newtonsoft.Json;
using Bogus;
using lmsservice_db.src.api.certificates.certificatesvalidation;
using lmsservice_db.src.api.certificates.certificatesmodel;
using FluentValidation;

namespace lmsservice_db.src.api.certificates.certificatesservice;

public class Certificates
{
    private readonly APIClient _client;

    public Certificates(APIClient client)
    {
        _client = client;
    }
    public Task<string> Create(CertificatesCreateModel certificates)
    {
        return _client.Request("/certificates", HttpMethod.Post, certificates);
    }

    public Task<string> GetById(string certificatesId)
    {
        return _client.Request($"/certificates/{certificatesId}", HttpMethod.Get);
    }

    public Task<string> Search(CertificatesSearchFilters searchFilters)
    {
        var queryParams = new List<string>();

        if (!string.IsNullOrEmpty(searchFilters.CertificateNumber)) queryParams.Add($"CertificateNumber={searchFilters.CertificateNumber}");
if (!string.IsNullOrEmpty(searchFilters.CertificateName)) queryParams.Add($"CertificateName={searchFilters.CertificateName}");
if (!string.IsNullOrEmpty(searchFilters.CertificateType)) queryParams.Add($"CertificateType={searchFilters.CertificateType}");
if (!string.IsNullOrEmpty(searchFilters.IssuedDate)) queryParams.Add($"IssuedDate={searchFilters.IssuedDate}");
if (searchFilters.ExpiryDate != null) queryParams.Add($"ExpiryDate={searchFilters.ExpiryDate.ToString("o")}");
if (!string.IsNullOrEmpty(searchFilters.CertificateUrl)) queryParams.Add($"CertificateUrl={searchFilters.CertificateUrl}");
if (searchFilters.FinalGrade != null) queryParams.Add($"FinalGrade={searchFilters.FinalGrade}");
if (searchFilters.CreditHours != null) queryParams.Add($"CreditHours={searchFilters.CreditHours}");
if (!string.IsNullOrEmpty(searchFilters.Skills)) queryParams.Add($"Skills={searchFilters.Skills}");
queryParams.Add($"IsVerified={searchFilters.IsVerified.ToString().ToLower()}");
if (!string.IsNullOrEmpty(searchFilters.VerificationUrl)) queryParams.Add($"VerificationUrl={searchFilters.VerificationUrl}");


        var queryString = string.Join("&", queryParams);
        return _client.Request($"/certificates/search?{queryString}", HttpMethod.Get);
    }

    public Task<string> Update(string certificatesId, CertificatesUpdateModel certificates)
    {
        return _client.Request($"/certificates/{certificatesId}", HttpMethod.Put, certificates);
    }

    public Task<string> Delete(string certificatesId)
    {
        return _client.Request($"/certificates/{certificatesId}", HttpMethod.Delete);
    }

    public async Task ExecuteCertificatesOperations()
    {
        var faker = new Faker("en");

        var createValidator = new CertificatesCreateModelValidator();
        var createNewCertificates = new CertificatesCreateModel
        {
                        UserId= "{{USER_ID}}",
            CertificateNumber= faker.Lorem.Sentence(3),
            CertificateName= faker.Name.FullName(),
            CertificateType= faker.Lorem.Sentence(3),
            IssuedDate= faker.Lorem.Sentence(3),
            ExpiryDate= faker.Date.Past(),
            CertificateUrl= faker.Internet.Url(),
            FinalGrade= faker.Random.Float(1, 100),
            CreditHours= faker.Random.Int(1, 10000),
            Skills= faker.Lorem.Sentence(3),
            IsVerified= faker.Random.Bool(),
            VerificationUrl= faker.Internet.Url(),
            IssuedBy= faker.Random.Guid().ToString(),

        };

        var createValidationResult = createValidator.Validate(createNewCertificates);
        if (!createValidationResult.IsValid)
        {
            foreach (var error in createValidationResult.Errors)
            {
                Console.WriteLine("Validation error:" + error.ErrorMessage);
            }
            return;
        }

        if (!ValidationHelper.TryValidateModel(createNewCertificates, out var validationResults))
        {
            foreach (var validationResult in validationResults)
            {
                Console.WriteLine(validationResult.ErrorMessage);
            }
            return;
        }

        var createResponse = await Create(createNewCertificates);
        dynamic createdCertificates = JsonConvert.DeserializeObject(createResponse);
        Console.WriteLine("Create: " + JsonConvert.SerializeObject(createdCertificates, Formatting.Indented));

        string certificatesId = createdCertificates.Data.id.ToString();

        var retrievedResponse = await GetById(certificatesId);
        dynamic retrieved = JsonConvert.DeserializeObject(retrievedResponse);
        Console.WriteLine("GetById: " + JsonConvert.SerializeObject(retrieved, Formatting.Indented));

        var searchValidator = new CertificatesSearchFiltersValidator();
        var searchFilters = new CertificatesSearchFilters
        {
                        CertificateNumber= createNewCertificates.CertificateNumber,
            CertificateName= createNewCertificates.CertificateName,
            CertificateType= createNewCertificates.CertificateType,
            IssuedDate= createNewCertificates.IssuedDate,
            ExpiryDate= createNewCertificates.ExpiryDate,
            CertificateUrl= createNewCertificates.CertificateUrl,
            FinalGrade= createNewCertificates.FinalGrade,
            CreditHours= createNewCertificates.CreditHours,
            Skills= createNewCertificates.Skills,
            IsVerified= createNewCertificates.IsVerified,
            VerificationUrl= createNewCertificates.VerificationUrl,

        };

        var searchValidationResult = searchValidator.Validate(searchFilters);
        if (!searchValidationResult.IsValid)
        {
            foreach (var error in searchValidationResult.Errors)
            {
                Console.WriteLine("Validation error:" + error.ErrorMessage);
            }
            return;
        }

        var searchResultsResponse = await Search(searchFilters);
        dynamic searchResults = JsonConvert.DeserializeObject(searchResultsResponse);
        Console.WriteLine("Search: " + JsonConvert.SerializeObject(searchResults, Formatting.Indented));

        var updateValidator = new CertificatesUpdateModelValidator();
        var updatedCertificates = new CertificatesUpdateModel
        {
                        UserId= "{{USER_ID}}",
            CertificateNumber= faker.Lorem.Sentence(3),
            CertificateName= faker.Name.FullName(),
            CertificateType= faker.Lorem.Sentence(3),
            IssuedDate= faker.Lorem.Sentence(3),
            ExpiryDate= faker.Date.Past(),
            CertificateUrl= faker.Internet.Url(),
            FinalGrade= faker.Random.Float(1, 100),
            CreditHours= faker.Random.Int(1, 10000),
            Skills= faker.Lorem.Sentence(3),
            IsVerified= faker.Random.Bool(),
            VerificationUrl= faker.Internet.Url(),
            IssuedBy= faker.Random.Guid().ToString(),

        };

        var updateValidationResult = updateValidator.Validate(updatedCertificates);
        if (!updateValidationResult.IsValid)
        {
            foreach (var error in updateValidationResult.Errors)
            {
                Console.WriteLine("Validation error:" + error.ErrorMessage);
            }
            return;
        }

        var updateResponse = await Update(certificatesId, updatedCertificates);
        dynamic updated = JsonConvert.DeserializeObject(updateResponse);
        Console.WriteLine("Update: " + JsonConvert.SerializeObject(updated, Formatting.Indented));

        var deleteResponse = await Delete(certificatesId);
        dynamic deleted = JsonConvert.DeserializeObject(deleteResponse);
        Console.WriteLine("Delete: " + JsonConvert.SerializeObject(deleted, Formatting.Indented));
    }
}
